### PR TITLE
Improve AnnotationDifferencer performance

### PR DESCRIPTION
**Pull request by tdclemens**
I am trying to use this plugin along with the [Checkstyle](https://wiki.jenkins-ci.org/display/JENKINS/Checkstyle+Plugin) plugin to parse rubocop warnings. The Checkstyle XML output output file is about 23 MB since I have a lot of rubocop warnings currently in my project.

Based on feedback from [this jenkins bug](https://issues.jenkins-ci.org/browse/JENKINS-2978) and [stack overflow](http://stackoverflow.com/questions/28671903/hashset-removeall-method-is-surprisingly-slow) I made this simple change, that should increase performance significantly for certain scenarios.

Basically `removeAll` works differently for HashSets or Collections because it calls `contains` in some edge cases.

### CODE REVIEW

File: src/main/java/hudson/plugins/analysis/core/AnnotationDifferencer.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
     private static Set<FileAnnotation> difference(
             final Set<FileAnnotation> target, final Set<FileAnnotation> other) {
         Set<FileAnnotation> difference = Sets.newHashSet(target);
-        difference.removeAll(other);
+        difference.removeAll(Sets.newHashSet(other));
         return difference;
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/core/AnnotationDifferencer.java]: But other already is a set?

Comment (tdclemens) [inline src/main/java/hudson/plugins/analysis/core/AnnotationDifferencer.java]: its a Set but not necessarily a HashSet. It could be an EnumSet or TreeSet or something. I traced one code path for the use of this method and found the Set is created in [getAnnotations()](https://github.com/jenkinsci/analysis-core-plugin/blob/master/src/main/java/hudson/plugins/analysis/util/model/AnnotationContainer.java#L368). ```annotations.values()``` is a collection. Im not sure if its implemented as a HashSet or something else. We want to use a HashSet because its implementation is quicker with ```removeAll```

Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/core/AnnotationDifferencer.java]: I see. Did you measure the improvements? Creating a new set is expensive, too.

Comment (tdclemens) [inline src/main/java/hudson/plugins/analysis/core/AnnotationDifferencer.java]: I haven't tested performance of this method yet -- just the performance of removeAll separately. Do you use any library for benchmarking? If not, I can try to find time to create a java program that tests the performance of this method with different sets and set sizes.

Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/core/AnnotationDifferencer.java]: I think it would be sufficient to run a test with a large amount of warnings and measure the system time. In getAnnotations an immutable set is returned.

I think we should not change code until we can make sure that it is really faster.
